# The Battle of Neighborhoods - Final Assignment

In this Final Assignment we will examine the Neighborhoods of Brussels in order to identify the best Neighborhoods for a young couple that recently moved to the Belgian capital. The Real Estate company recently added a additional service that offers their clients a tailored service including a detailed report on the best Neighborhoods based on the preferences of the client.

## 1. Introduction

A young couple recently moved to the Belgian capital for work and are wondering which of the Neighborhoods of Brussels will be the best to live in. Their interests include:  
1. The outdoors
2. Italian Restaurants
3. Cultural activities

They just saw an advertisment for a tailored service from a Real Estate company that will help them identify the top 3 Neighborhoods to choose based on their interests. Therefore, they contacted this Real Estate company and transmitted their interests.  
The young couple both are in their 30s and work full-time and don't know Brussels at all. 

The Real Estate company analyzes their interests and determines that the best way to come up with the best Neighborhoods they will need the coordinates of Brussels and combine this with the Foursquare location data. 

## 2. Data

In order to determine the best Neighborhoods for the young couple the Real Estate company first analyzes how Brussels is divided. Brussels is the capital of Belgium and is divided in 19 municipalities, whith a defined Postal Code.  
Inside the municipalities there are different Neighborhoods, however, as they are only localities and don't have very defined limits nor Postal Codes, the Real Estate company decided to restrict their recommendation to one of the 19 municipalities in Brussels.  

To find a list of the 19 Brussels Municipalities the Real Estate company found the Wikipedia page that has a table that lists the 19 municipalities with their corresponding Postal Codes.

In order to get the Longitutde and Latitude coordinates of the 19 different municipalities, the Real Estate company has access to the Open Data website of Wallonia and Brussels that hosts a CSV file listing all Belgian Postal Codes with the names of the municipalities as well as their Longitude and Latitudes.

Once we have the latitude and longitude of the identified municipalities of Brussels we can then use Foursquare's location data to identify the type of venues in the different municipalities. This could then give us a good idea of which municipality would match the criteria's given by the young couple.

### 2.1 Scrapping and Cleaning the Wikipedia page table

To start let us scrape the Wikipedia page that hosts the Table of the different 19 municipalities of Brussels.  

The first step is to install and import the different modules we will need to scrape and clean the Table. For this operation we will need Pandas and Numpy.

In [1]:
import pandas as pd
import numpy as np

To scrape the Wikiwand website we will use the Pandas read_html method that returns HTML tables into a list of DataFrame objects.  

The URL of the Wikipedia page that contains the Table is: "https://fr.wikipedia.org/wiki/R%C3%A9gion_de_Bruxelles-Capitale"

In [2]:
url = 'https://fr.wikipedia.org/wiki/R%C3%A9gion_de_Bruxelles-Capitale'

df_list = pd.read_html(url)

len(df_list)

16

We can see that read_html identified 16 different HTML tables. From navigating to the Wikipedia page we can identify that the table we are interested in is the second HTML table present. Therefore, let's only look at the second DataFrame object.

In [3]:
brussels_list = df_list[1]

brussels_list

Repèresur lacarte                                            Commune  \
0                 1.0                                         Anderlecht   
1                 2.0                                          Auderghem   
2                 3.0                              Berchem-Sainte-Agathe   
3                 4.0  Ville de Bruxelles : Bruxelles Laeken Neder-Ov...   
4                 5.0                                          Etterbeek   
5                 6.0                                              Evere   
6                 7.0                                             Forest   
7                 8.0                                          Ganshoren   
8                 9.0                                            Ixelles   
9                10.0                                              Jette   
10               11.0                                         Koekelberg   
11               12.0                               Molenbeek-Saint-Jean   
12               13.0                                       Saint-Gilles   
13               14.0                              Saint-Josse-ten-Noode   
14               15.0                                         Schaerbeek   
15               16.0                                              Uccle   
16               17.0                                Watermael-Boitsfort   
17               18.0                               Woluwe-Saint-Lambert   
18               19.0                                Woluwe-Saint-Pierre   
19                NaN                                             Totaux   

                                    Code postal  Superficie(en km2)  \
0                                          1070                1774   
1                                          1160                 903   
2                                          1082                 295   
3   Plusieurs : 1000, 1040, 1050 1020 1120 1130                3261   
4                                          1040                 315   
5                                          1140                 507   
6                                          1190                 625   
7                                          1083                 246   
8                                          1050                 634   
9                                          1090                 504   
10                                         1081                 117   
11                                         1080                 589   
12                                         1060                 252   
13                                         1210                 114   
14                                         1030                 814   
15                                         1180                2291   
16                                         1170                1293   
17                                         1200                 722   
18                                         1150                 885   
19                                          NaN               16136   

   Population(au 1/1/2009) Population(au 1/1/2010) Population(au 1/1/2016)  \
0                  101 371                 104 647                 116 494   
1                   30 456                  30 811                  33 085   
2                   21 669                  22 185                  24 176   
3                  153 377                 157 673                 176 512   
4                   43 512                  44 352                  47 023   
5                   35 372                  35 803                  39 439   
6                   49 757                  50 258                  55 386   
7                   22 160                  22 589                  24 128   
8                   80 312                  80 183                  85 274   
9                   45 637                  46 818                  51 387   
10                  19 380                  19 812       

From the above DataFrame we can see that the table includes a lot of information that we don't need for the moment. Let's only keep the relevant columns, "Commune" and "Code postal" which translates into "municipalities" and "Postal Code". 

In [4]:
brussels_list =  brussels_list [['Commune', 'Code postal']]

brussels_list

Commune  \
0                                          Anderlecht   
1                                           Auderghem   
2                               Berchem-Sainte-Agathe   
3   Ville de Bruxelles : Bruxelles Laeken Neder-Ov...   
4                                           Etterbeek   
5                                               Evere   
6                                              Forest   
7                                           Ganshoren   
8                                             Ixelles   
9                                               Jette   
10                                         Koekelberg   
11                               Molenbeek-Saint-Jean   
12                                       Saint-Gilles   
13                              Saint-Josse-ten-Noode   
14                                         Schaerbeek   
15                                              Uccle   
16                                Watermael-Boitsfort   
17                               Woluwe-Saint-Lambert   
18                                Woluwe-Saint-Pierre   
19                                             Totaux   

                                    Code postal  
0                                          1070  
1                                          1160  
2                                          1082  
3   Plusieurs : 1000, 1040, 1050 1020 1120 1130  
4                                          1040  
5                                          1140  
6                                          1190  
7                                          1083  
8                                          1050  
9                                          1090  
10                                         1081  
11                                         1080  
12                                         1060  
13                                         1210  
14                                         1030  
15                                         1180  
16                                         1170  
17                                         1200  
18                                         1150  
19                                          NaN

The table looks clean. However, we can see that the fourth row has more than one name and more than one postal code. That's because the central municipality of Brussels includes, aside of itself 3 other regions that formerly were seperate municipalities but are now attached to the municipality of Brussels. As they were formerly municipalities these 3 regions have distinct postal codes.  

Therefore, we will add these regions to the complete DataFrame.  

First let's have a look to the fourth row in detail.

In [5]:
city_brussels_PostalCode = brussels_list.iloc[3]

city_brussels_PostalCode

Commune        Ville de Bruxelles : Bruxelles Laeken Neder-Ov...
Code postal          Plusieurs : 1000, 1040, 1050 1020 1120 1130
Name: 3, dtype: object

In [6]:
city_brussels_name = city_brussels_PostalCode['Commune']

print(city_brussels_name)

Ville de Bruxelles : Bruxelles Laeken Neder-Over-Heembeek Haren


We can see that the row includes 4 distinct names. Namely: Bruxelles, Laeken, Neder-Over-Heembeek, and Haren. Therefore, let's create a list that includes these 4 municipalities names. 

In [7]:
city_brussels_name_list = city_brussels_name.split(" ")
city_brussels_name_list

['Ville',
 'de',
 'Bruxelles\xa0:',
 'Bruxelles',
 'Laeken',
 'Neder-Over-Heembeek',
 'Haren']

In [8]:
city_brussels_name_lists = city_brussels_name_list[3:7]

city_brussels_name_lists

['Bruxelles', 'Laeken', 'Neder-Over-Heembeek', 'Haren']

Let's have a look at the postal codes in this row. 

In [9]:
city_brussels_PostalCode_lists = city_brussels_PostalCode['Code postal'].split(" ")

city_brussels_PostalCode_lists

['Plusieurs\xa0:', '1000,', '1040,', '1050', '1020', '1120', '1130']

We can see that there are more postal codes than names. That's because "Bruxelles" had a sub-list of postal codes; 1000, 1040, and 1050. 

In [10]:
city_brussels_PostalCode_lists = city_brussels_PostalCode_lists[1:7]

city_brussels_PostalCode_lists

['1000,', '1040,', '1050', '1020', '1120', '1130']

Upon further inspection we can see that two of these postal codes; 1040 and 1050, are already present in the table. Therefore, we will only keep 1000 for Bruxelles.

In [11]:
del city_brussels_PostalCode_lists[1:3]

The last step is to remove the comma that appears after 1000.

In [12]:
city_brussels_PostalCode_lists = [city_brussels_PostalCode_lists[0].replace('1000,', '1000'), city_brussels_PostalCode_lists[1],city_brussels_PostalCode_lists[2], city_brussels_PostalCode_lists[3]]

city_brussels_PostalCode_lists

['1000', '1020', '1120', '1130']

With both list clean we are ready to merge them and create a sub-DataFrame.

In [13]:
brussels_city_df = pd.DataFrame({'Commune':city_brussels_name_lists, 'Code postal': city_brussels_PostalCode_lists})

brussels_city_df

Commune Code postal
0            Bruxelles        1000
1               Laeken        1020
2  Neder-Over-Heembeek        1120
3                Haren        1130

Now its time to fuse the two DataFrames together.

In [14]:
full_df = brussels_list.append(brussels_city_df, ignore_index=True)

full_df

Commune  \
0                                          Anderlecht   
1                                           Auderghem   
2                               Berchem-Sainte-Agathe   
3   Ville de Bruxelles : Bruxelles Laeken Neder-Ov...   
4                                           Etterbeek   
5                                               Evere   
6                                              Forest   
7                                           Ganshoren   
8                                             Ixelles   
9                                               Jette   
10                                         Koekelberg   
11                               Molenbeek-Saint-Jean   
12                                       Saint-Gilles   
13                              Saint-Josse-ten-Noode   
14                                         Schaerbeek   
15                                              Uccle   
16                                Watermael-Boitsfort   
17                               Woluwe-Saint-Lambert   
18                                Woluwe-Saint-Pierre   
19                                             Totaux   
20                                          Bruxelles   
21                                             Laeken   
22                                Neder-Over-Heembeek   
23                                              Haren   

                                    Code postal  
0                                          1070  
1                                          1160  
2                                          1082  
3   Plusieurs : 1000, 1040, 1050 1020 1120 1130  
4                                          1040  
5                                          1140  
6                                          1190  
7                                          1083  
8                                          1050  
9                                          1090  
10                                         1081  
11                                         1080  
12                                         1060  
13                                         1210  
14                                         1030  
15                                         1180  
16                                         1170  
17                                         1200  
18                                         1150  
19                                          NaN  
20                                         1000  
21                                         1020  
22                                         1120  
23                                         1130

We can see that this Dataframe still includes the 'faulty' row. Furthermore, it includes a Total row that has a *NaN* value. Therefore, let's drop these two rows.

In [15]:
full_df.drop([3, 19], inplace=True)

Lastly, let's sort the DataFrame by increasing number of postal code.

In [16]:
full_df.sort_values(by=['Code postal'], ignore_index=True)

Commune Code postal
0               Bruxelles        1000
1                  Laeken        1020
2              Schaerbeek        1030
3               Etterbeek        1040
4                 Ixelles        1050
5            Saint-Gilles        1060
6              Anderlecht        1070
7    Molenbeek-Saint-Jean        1080
8              Koekelberg        1081
9   Berchem-Sainte-Agathe        1082
10              Ganshoren        1083
11                  Jette        1090
12    Neder-Over-Heembeek        1120
13                  Haren        1130
14                  Evere        1140
15    Woluwe-Saint-Pierre        1150
16              Auderghem        1160
17    Watermael-Boitsfort        1170
18                  Uccle        1180
19                 Forest        1190
20   Woluwe-Saint-Lambert        1200
21  Saint-Josse-ten-Noode        1210

In [17]:
full_df.shape

(22, 2)

We see that the DataFrame now includes 21 postal codes, which corresponds to the 19 municipalities of Brussels plus the two added sub-regions of the Brussels municipality.

## 2.2 Adding the Latitude and Longitude data to the municipalities.

In this section we will explore how to add the Latitude and Longitude coordinates to the municipalities using the postal codes as an index. 

The CSV used is openly accessible from the Open Data Wallonie-Bruxelles platform. [Link](https://www.odwb.be/explore/dataset/code-postaux-belge/information/)

In [18]:
csv = 'https://www.odwb.be/explore/dataset/code-postaux-belge/download/?format=csv&timezone=Europe/Berlin&lang=fr&use_labels_for_header=true&csv_separator=%3B'

lat_long = pd.read_csv(csv, sep=';')

lat_long.head()

Code             Localite  Longitude   Latitude               Coordonnees  \
0  1050              Ixelles   4.381571  50.822285      50.8222854,4.3815707   
1  1081           Koekelberg   4.325708  50.862263      50.8622633,4.3257084   
2  1130                Haren   4.412571  50.891966  50.8919661,4.41257087324   
3  1150  Woluwe-Saint-Pierre   4.443297  50.829243      50.8292429,4.4432967   
4  1315              Incourt   4.798944  50.691800       50.6918004,4.798944   

   Geom  
0   NaN  
1   NaN  
2   NaN  
3   NaN  
4   NaN

In [19]:
lat_long = lat_long[['Code', 'Latitude', 'Longitude']]

In [20]:
lat_long.rename(columns = {'Code': 'Code postal'}, inplace = True)

In [28]:
brussels_df = full_df.merge(lat_long,on=['Code postal'])

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat

In [29]:
full_df.dtypes

Commune        object
Code postal    object
dtype: object

In [30]:
lat_long.dtypes

Code postal      int64
Latitude       float64
Longitude      float64
dtype: object

In [36]:
full_df = full_df.astype({'Code postal':'int64'})

In [37]:
full_df.dtypes

Commune        object
Code postal     int64
dtype: object

In [38]:
brussels_df = full_df.merge(lat_long,on=['Code postal'])

In [40]:
brussels_df

Commune  Code postal   Latitude  Longitude
0              Anderlecht         1070  50.838141   4.312340
1               Auderghem         1160  50.815657   4.433139
2   Berchem-Sainte-Agathe         1082  50.863984   4.292702
3               Etterbeek         1040  50.836851   4.389510
4                   Evere         1140  50.870452   4.402160
5                  Forest         1190  50.809143   4.317751
6               Ganshoren         1083  50.871240   4.317510
7                 Ixelles         1050  50.822285   4.381571
8                   Jette         1090  50.877763   4.326090
9              Koekelberg         1081  50.862263   4.325708
10   Molenbeek-Saint-Jean         1080  50.854355   4.322778
11           Saint-Gilles         1060  50.826741   4.345668
12  Saint-Josse-ten-Noode         1210  50.853074   4.372336
13             Schaerbeek         1030  50.867604   4.373712
14                  Uccle         1180  50.801820   4.337235
15    Watermael-Boitsfort         1170  50.799394   4.415818
16   Woluwe-Saint-Lambert         1200  50.846693   4.428484
17    Woluwe-Saint-Pierre         1150  50.829243   4.443297
18              Bruxelles         1000  50.846557   4.351697
19                 Laeken         1020  50.883392   4.348713
20    Neder-Over-Heembeek         1120  50.897796   4.390489
21                  Haren         1130  50.891966   4.412571

## 2.3 Visualization of the municipalities

In this section we will visualize the identified municipalities on a Map of Brussels.  

First lets install and import the need packages and modules.

In [41]:
import folium #package used to visualize a map

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import numpy as np # library to handle data in a vectorized manner

print('packages installed and imported')

import json

packages installed and imported


Let's use the Nominatim geolocator to find out the latitude and longitude of Brussels. 

In [42]:
address = 'Brussels'

geolocator = Nominatim(user_agent="br_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Brussels are: {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Brussels are: 50.8465573, 4.351697.


We'll use Brussels latitude and longtiude to center the Folium map and add markers to the identified municipalities on the map. 

In [43]:
map_brussels = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, municipality in zip(brussels_df['Latitude'], brussels_df['Longitude'], brussels_df['Commune']):
    label = '{}'.format(municipality)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_brussels)  
    
map_brussels